skimage/segmentation/_join.py

```diff
@@ -116,26 +115,195 @@ def relabel_sequential(label_field, offset=1):
     >>> relab
     array([5, 5, 6, 6, 7, 9, 8])
     """
-    offset = int(offset)
     if offset <= 0:
         raise ValueError("Offset must be strictly positive.")
     if np.min(label_field) < 0:
         raise ValueError("Cannot relabel array that contains negative values.")
-    max_label = int(label_field.max()) # Ensure max_label is an integer
-    if not np.issubdtype(label_field.dtype, np.integer):
-        new_type = np.min_scalar_type(max_label)
-        label_field = label_field.astype(new_type)
-    labels = np.unique(label_field)
-    labels0 = labels[labels != 0]
-    new_max_label = offset - 1 + len(labels0)
-    new_labels0 = np.arange(offset, new_max_label + 1)
-    output_type = label_field.dtype
-    required_type = np.min_scalar_type(new_max_label)
-    if np.dtype(required_type).itemsize > np.dtype(label_field.dtype).itemsize:
+    offset = int(offset)
+    in_vals = np.unique(label_field)
```

Comment:
It seems that `out_array` is almost accessible via the `return_inverse ` of `np.unique`:
```python
>>> import numpy as np
>>> from skimage.segmentation import relabel_sequential
>>> 
>>> offset = 2
>>> label_field = np.array([[1, 1, 5, 5, 8, 99, 42],
...                         [1, 0, 0, 5, 99, 99, 42]])
>>> relab, fw, inv = relabel_sequential(label_field, offset=offset)
>>> 
>>> in_vals, out_array = np.unique(label_field, return_inverse=True)
>>> 
>>> out_array = out_array.reshape(label_field.shape) + offset - (0 in in_vals)
>>> out_array[label_field == 0] = 0
>>> relab
array([[2, 2, 3, 3, 4, 6, 5],
       [2, 0, 0, 3, 6, 6, 5]])
>>> out_array
array([[2, 2, 3, 3, 4, 6, 5],
       [2, 0, 0, 3, 6, 6, 5]])
```
These [3 lines](https://github.com/scikit-image/scikit-image/pull/4612/files#diff-09660f97c6b4c3dc5d143d3be12e8aaaR150-R152) are no more needed.

Reply:
@rfezzani Fascinating!!!

However, if we are building the map arrays anyway, this is actually faster than the overhead of running np.unique with `return_inverse=True`:

```python
In [3]: %timeit np.unique(seg)                                                     
3.83 ms ± 229 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
In [4]: %timeit np.unique(seg, return_inverse=True)                                
7.12 ms ± 399 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
In [5]: %timeit j.relabel_sequential(seg)                                          
6.2 ms ± 36.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
In [6]: reseg, fw, inv = j.relabel_sequential(seg)                                 
In [7]: %timeit fw[seg]                                                            
2.33 ms ± 152 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
In [8]: %timeit np.empty_like(seg)                                                 
547 ns ± 6.14 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Having said that, this would have been a useful refactor before this Cython code. :sweat_smile: